File: Java_version/app/src/beta/java/com/termux/wayland/beta/data/AppRepository.java
package com.termux.wayland.beta.data;

import android.content.Context;
import android.os.AsyncTask;

import com.termux.wayland.beta.scanner.DesktopFileScanner;
import com.termux.wayland.beta.scanner.DpkgChecker;
import com.termux.wayland.beta.scanner.JsonAppScanner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Point d'entrée unique pour obtenir la liste des apps Linux disponibles.
 *
 * Pipeline complet :
 *   1. Charger dpkg (liste des paquets installés)
 *   2. Lire apps.json (base curatée)
 *   3. Lire apps_user.json (overrides utilisateur)
 *   4. Merger JSON (USER > DATABASE)
 *   5. Scanner .desktop (fallback)
 *   6. Fusionner tout
 *   7. Filtrer : garder seulement les apps installées
 *   8. Dédupliquer par id
 *   9. Trier par nom
 */
public class AppRepository {

    /**
     * Callback appelé quand la liste est prête.
     */
    public interface OnAppsLoadedListener {
        void onAppsLoaded(List<LinuxApp> apps);
        void onError(Exception e);
    }

    // Singleton
    private static AppRepository instance;

    private AppRepository() {}

    public static synchronized AppRepository getInstance() {
        if (instance == null) {
            instance = new AppRepository();
        }
        return instance;
    }

    /**
     * Charge les apps de façon asynchrone.
     * Résultat livré via le listener sur le thread principal.
     *
     * @param context Android Context
     * @param includeDatabaseOnly si true, ignore le scan .desktop (plus rapide)
     * @param listener callback résultat
     */
    public void getInstalledApps(
            Context context,
            boolean includeDatabaseOnly,
            OnAppsLoadedListener listener
    ) {
        new LoadAppsTask(context, includeDatabaseOnly, listener).execute();
    }

    /**
     * Surcharge rapide (inclut .desktop scan).
     */
    public void getInstalledApps(Context context, OnAppsLoadedListener listener) {
        getInstalledApps(context, false, listener);
    }

    /**
     * Force un rechargement complet (utile après installation d'un paquet).
     */
    public void refresh(Context context, OnAppsLoadedListener listener) {
        DpkgChecker.getInstance().refresh();
        getInstalledApps(context, listener);
    }

    // ─── AsyncTask interne ────────────────────────────────────────────────────

    private static class LoadAppsTask extends AsyncTask<Void, Void, List<LinuxApp>> {

        private final Context context;
        private final boolean includeDatabaseOnly;
        private final OnAppsLoadedListener listener;
        private Exception error;

        LoadAppsTask(Context context, boolean includeDatabaseOnly, OnAppsLoadedListener listener) {
            this.context              = context.getApplicationContext();
            this.includeDatabaseOnly  = includeDatabaseOnly;
            this.listener             = listener;
        }

        @Override
        protected List<LinuxApp> doInBackground(Void... voids) {
            try {
                // ── Étape 1 : initialisation ──────────────────────────────────
                JsonAppScanner.getInstance().initUserFileIfNeeded(context);
                DpkgChecker.getInstance().loadInstalledPackages();

                // ── Étapes 2-4 : sources JSON ─────────────────────────────────
                List<LinuxApp> jsonApps = JsonAppScanner.getInstance().loadApps(context);

                // ── Étape 5 : scan .desktop (optionnel) ───────────────────────
                List<LinuxApp> desktopApps = new ArrayList<>();
                if (!includeDatabaseOnly) {
                    desktopApps = DesktopFileScanner.getInstance().scanApps();
                }

                // ── Étape 6 : fusion globale ──────────────────────────────────
                List<LinuxApp> allApps = mergeWithDesktop(jsonApps, desktopApps);

                // ── Étape 7 : filtre dpkg ─────────────────────────────────────
                DpkgChecker dpkg = DpkgChecker.getInstance();
                List<LinuxApp> filtered = new ArrayList<>();
                for (LinuxApp app : allApps) {
                    if (app.getSource() == LinuxApp.AppSource.DESKTOP) {
                        filtered.add(app); // pas de vérif possible
                    } else if (dpkg.isInstalled(app.getPackageName())) {
                        filtered.add(app);
                    }
                }

                // ── Étapes 8-9 : déduplique + tri ────────────────────────────
                return deduplicateAndSort(filtered);

            } catch (Exception e) {
                this.error = e;
                return new ArrayList<>();
            }
        }

        @Override
        protected void onPostExecute(List<LinuxApp> apps) {
            if (error != null) {
                listener.onError(error);
            } else {
                listener.onAppsLoaded(apps);
            }
        }
    }

    // ─── Helpers statiques ────────────────────────────────────────────────────

    /**
     * Fusionne apps JSON + apps .desktop.
     * Les ids JSON ont priorité sur .desktop.
     */
    private static List<LinuxApp> mergeWithDesktop(
            List<LinuxApp> jsonApps,
            List<LinuxApp> desktopApps
    ) {
        Set<String> knownIds = new HashSet<>();
        for (LinuxApp app : jsonApps) {
            knownIds.add(app.getId());
        }

        List<LinuxApp> result = new ArrayList<>(jsonApps);
        for (LinuxApp app : desktopApps) {
            if (!knownIds.contains(app.getId())) {
                result.add(app);
            }
        }
        return result;
    }

    /**
     * Supprime les doublons (même id) et trie par nom alphabétique.
     */
    private static List<LinuxApp> deduplicateAndSort(List<LinuxApp> apps) {
        Set<String> seen = new HashSet<>();
        List<LinuxApp> unique = new ArrayList<>();
        for (LinuxApp app : apps) {
            if (seen.add(app.getId())) {
                unique.add(app);
            }
        }
        Collections.sort(unique, (a, b) ->
            a.getName().toLowerCase().compareTo(b.getName().toLowerCase())
        );
        return unique;
    }
}

File: Kotlin_version/app/src/main/java/com/termux/wayland/data/AppRepository.kt
package com.termux.wayland.data

import android.content.Context
import com.termux.wayland.scanner.DesktopFileScanner
import com.termux.wayland.scanner.DpkgChecker
import com.termux.wayland.scanner.JsonAppScanner
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Point d'entrée unique pour obtenir la liste des apps Linux disponibles.
 *
 * Pipeline complet :
 *   1. Charger dpkg (liste des paquets installés)
 *   2. Lire apps.json (base curatée)
 *   3. Lire apps_user.json (overrides utilisateur)
 *   4. Merger JSON (USER > DATABASE)
 *   5. Scanner .desktop (fallback pour apps inconnues)
 *   6. Fusionner tout
 *   7. Filtrer : ne garder que les apps installées (via dpkg)
 *   8. Dédupliquer par id
 *   9. Trier par nom
 */
object AppRepository {

    /**
     * Charge et retourne toutes les apps installées.
     * Suspendu — à appeler depuis une coroutine.
     *
     * @param context Android Context
     * @param includeDatabaseOnly si true, ignore le scan .desktop (plus rapide)
     */
    suspend fun getInstalledApps(
        context: Context,
        includeDatabaseOnly: Boolean = false
    ): List<LinuxApp> = withContext(Dispatchers.IO) {

        // ── Étape 1 : initialisation ──────────────────────────────────────────
        JsonAppScanner.initUserFileIfNeeded(context)
        DpkgChecker.loadInstalledPackages()

        // ── Étape 2-4 : sources JSON ──────────────────────────────────────────
        val jsonApps = JsonAppScanner.loadApps(context)

        // ── Étape 5 : scan .desktop (optionnel) ───────────────────────────────
        val desktopApps = if (!includeDatabaseOnly) {
            DesktopFileScanner.scanApps()
        } else {
            emptyList()
        }

        // ── Étape 6 : fusion globale ──────────────────────────────────────────
        // Les entrées JSON (DATABASE/USER) ont priorité sur DESKTOP
        // Si un .desktop a le même id qu'une entrée JSON, on garde le JSON
        val allApps = mergeWithDesktop(jsonApps, desktopApps)

        // ── Étape 7 : filtre dpkg ─────────────────────────────────────────────
        // Pour les apps DESKTOP, on ne connaît pas le vrai package → on les garde
        // si leur exec est dans le PATH (on ne peut pas vérifier avec dpkg)
        val filtered = allApps.filter { app ->
            when (app.source) {
                AppSource.DESKTOP  -> true          // pas de paquet connu → on garde
                AppSource.DATABASE,
                AppSource.USER     -> DpkgChecker.isInstalled(app.packageName)
            }
        }

        // ── Étapes 8-9 : déduplique + tri ────────────────────────────────────
        filtered
            .distinctBy { it.id }
            .sortedBy { it.name.lowercase() }
    }

    /**
     * Fusionne les apps JSON avec les apps .desktop.
     * Les ids JSON ont priorité : si un .desktop a le même id, il est ignoré.
     */
    private fun mergeWithDesktop(
        jsonApps: List<LinuxApp>,
        desktopApps: List<LinuxApp>
    ): List<LinuxApp> {
        val knownIds = jsonApps.map { it.id }.toSet()
        val newDesktopApps = desktopApps.filter { it.id !in knownIds }
        return jsonApps + newDesktopApps
    }

    /**
     * Force un rechargement complet (utile après installation d'un nouveau paquet).
     */
    suspend fun refresh(context: Context): List<LinuxApp> {
        DpkgChecker.refresh()
        return getInstalledApps(context)
    }

    /**
     * Retourne uniquement les apps d'une catégorie donnée.
     */
    suspend fun getAppsByCategory(
        context: Context,
        category: String
    ): List<LinuxApp> {
        return getInstalledApps(context).filter { it.category == category }
    }

    /**
     * Retourne les catégories distinctes ayant au moins une app installée.
     */
    suspend fun getAvailableCategories(context: Context): List<String> {
        return getInstalledApps(context)
            .map { it.category }
            .distinct()
            .sorted()
    }
}
